### PR TITLE
Feature: Tested utm tracking snippet

### PR DESCRIPTION
**Pull request by tday**
Files to review:
`src/*`
`test/*`
`package.json`

Original code:
```
<script src="//front.embarkvet.com/js/lib/xdomain-v1.0.8/xdomain_cookie.min.js"></script>
<script>
  (function () {
    var TRACKING_ENDPOINT = 'https://my.embarkvet.com/utm-attribution';

    function getReferrerSource () {
      // Organic marketing attribution spec: https://docs.google.com/document/d/19CeUT7pDuWGk6hS3HlMOMnu1kWfCLmcNNvlnCaaLfus/edit#
      if (!document.referrer || document.referrer === '') { return; }
      var sources = {
        google: /^(www\.|)google\.([a-z]+|com?\.[a-z]+)$/,
        yahoo: /^search\.yahoo\.com$/,
        duckduckgo: /^duckduckgo\.com$/,
        bing: /www\.bing\.com$/,
        baidu: /^www\.baidu\.com$/,
        sogou: /^.*\.sogou\.com$/,
        yandex: /^(www\.|)yandex\.([a-z]+|com?\.[a-z]+)$/,
        naver: /^search\.naver\.com$/,
        aol: /^r\.search\.aol\.com$/,
        facebook: /^.*\.facebook\.com$/,
        instagram: /^.*\.instagram\.com$/,
        twitter: /^(t\.co|(www\.|)twitter\.com)$/,
        pinterest: /^www\.pinterest\.([a-z]+|com?\.[a-z]+)$/,
        reddit: /^www\.reddit\.com$/,
      };

      for (var channel in sources) {
        // check if the property/key is defined in the object itself, not in parent
        if (sources.hasOwnProperty(channel)) {
          var domain = sources[channel]
          if (domain.test(document.referrer.toLowerCase().split('/')[2])) {
            //console.log('UTM Manufacturing Domain Match Found!');
            //console.log(channel + ':  ' + domain + '  ==  ' + document.referrer.toLowerCase().split('/')[2]);
            return channel;
          }
        }
      }
      //console.log('No UTM Manufacturing Domain Match Found: ' + document.referrer.toLowerCase().split('/')[2]);

    };

    window.embark = window.embark || {};
    var onPageWithUtms = location.search.toLowerCase().indexOf('utm_') !== -1;
    var onOrderConfirmationPage = (location.hostname === 'shop.embarkvet.com') && window.Shopify && window.Shopify.checkout && window.Shopify.checkout.order_id;
    var justActivatedSwab = window.embark.justActivatedSwab;
    var isNewAccountCreation = window.embark.isNewAccountCreation;
    var isUtmTrackingSubmissionEvent = onOrderConfirmationPage || justActivatedSwab || isNewAccountCreation;

    var referrerSource
    if (!onPageWithUtms) {
      referrerSource = getReferrerSource()
    }

    var queryString
    if (onPageWithUtms) {
      // cookie values must not contain '='
      queryString = location.search.toLowerCase().replace(/=/g, '-->>');
    } else if (referrerSource) {
      queryString = '?utm_medium-->>organic&utm_source-->>' + referrerSource + '&utm_content-->>' + document.location.href
    }

    if (!queryString && !isUtmTrackingSubmissionEvent) {
      return;
    }

    window.embark.utmCookieJar = xDomainCookie('//front.embarkvet.com/js/lib/xdomain-v1.0.8', 'embark', true);

    if (queryString) {
      window.embark.utmCookieJar.get('utmFirst', function (utmFirst) {
        if (!utmFirst) {
          window.embark.utmCookieJar.set('utmFirst', JSON.stringify({
            at: new Date(),
            q: queryString
          }));
        }
      })
      window.embark.utmCookieJar.set('utmLast', JSON.stringify({
        at: new Date(),
        q: queryString
      }));
    }

    if (isUtmTrackingSubmissionEvent) {
      window.embark.utmCookieJar.get('utmFirst', function (utmFirst) {
        window.embark.utmCookieJar.get('utmLast', function (utmLast) {
          if (!utmFirst || !utmLast) {
            return
          }


          var data = {
            utmFirst: JSON.parse(utmFirst),
            utmLast: JSON.parse(utmLast)
          };
          var url
          if (onOrderConfirmationPage) {
            data.shopifyOrderId = '' + window.Shopify.checkout.order_id;
            url = TRACKING_ENDPOINT + '/shopify-order-confirmation';
          } else if (justActivatedSwab) {
            data.swab = '' + justActivatedSwab;
            url = TRACKING_ENDPOINT + '/swab-activated';
          } else if (isNewAccountCreation) {
            url = TRACKING_ENDPOINT + '/account-created';
          }
          var client = new XMLHttpRequest();
          client.open('POST', url);
          client.setRequestHeader('Content-Type', 'text/plain; charset=UTF-8');
          client.send(JSON.stringify(data));
          //console.log('UTM event sent using XMLHttpRequest', data);
        })
      })
    }
  })()
</script>
```

### CODE REVIEW

**Comment by tday** [comment]
[Make utm-attribution script production grade](https://trello.com/c/9huheL8k/911-make-utm-attribution-script-production-grade)

File: package.json [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  "scripts": {
+    "test": "jest",
+    "lint": "eslint . --ext .js --cache"
+  },
+  "author": "",
+  "license": "ISC",
+  "repository": "https://github.com/ProjectAussie/utm-client",
+  "dependencies": {},
+  "devDependencies": {
+    "@babel/core": "^7.7.2",
```
Comment (dlaudate-embark) [inline package.json]: Lock versions

File: package.json [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    "@babel/preset-env": "^7.7.1",
+    "@types/jest": "24.0.23",
+    "babel-jest": "^24.9.0",
+    "eslint": "6.6.0",
+    "eslint-config-prettier": "6.7.0",
+    "eslint-plugin-prettier": "3.1.1",
+    "jest": "24.9.0",
+    "prettier": "1.19.1",
+    "xdomain-cookies": "1.0.8"
```
Comment (DaveLak) [inline package.json]: Shouldn't this be in `dependencies` because the code in production relies on it?

File: package.json [hunk 3]
```diff
@@ -0,0 +1,25 @@
+{
+  "name": "utm-client",
+  "version": "1.0.0",
+  "description": "UTM attribution client",
+  "scripts": {
+    "test": "jest",
+    "lint": "eslint . --ext .js --cache"
```
Comment (DaveLak) [inline package.json]: May as well add a fix script to?
```suggestion
    "lint": "eslint . --ext .js --cache",
    "lint:fix": "eslint . --ext .js --fix"
```

File: babel.config.js [hunk 4]
```diff
@@ -0,0 +1,12 @@
+module.exports = {
+  presets: [
+    [
+      '@babel/preset-env',
+      {
+        targets: {
+          node: 'current',
```
Comment (DaveLak) [inline babel.config.js]: Is the plan to incorporate this into myev as a dependency? If not should't this be targeting browsers?

File: package.json (hunk 2 above)
Comment (tday) [reply package.json]: I'm going to attempt to not bundle it with our script. Unless you think I should bundle it?

Comment (DaveLak) [reply package.json]: And load it via CDN or something?

File: src/attribution.js [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    hostname === 'shop.embarkvet.com' && window.Shopify && window.Shopify.checkout && window.Shopify.checkout.order_id;
+  const justActivatedSwab = window.embark.justActivatedSwab;
+  const isNewAccountCreation = window.embark.isNewAccountCreation;
+  const isUtmTrackingSubmissionEvent = onOrderConfirmationPage || justActivatedSwab || isNewAccountCreation;
+  if (!isUtmTrackingSubmissionEvent) {
+    return;
+  }
+
+  const utmCookieJar = loadAndSaveCookieJar();
+  const utmFirst = await utmCookieJar.get('utmFirst');
```
Comment (DaveLak) [inline src/attribution.js]: You might run into trouble using `async`/`await` in code intended for the browser. As far as I'm aware Babel can't transpile it without [an extra library](https://babeljs.io/docs/en/babel-runtime) and pulling that library in may result in a bundle much larger than this project can justify.

File: package.json (hunk 2 above)
Comment (tday) [reply package.json]: yeah, it is already in our CDN

File: src/attribution.js (hunk 5 above)
Comment (tday) [reply src/attribution.js]: fair point. I can revert to callbacks or promises if it balloons. I think my refactor makes it simple to use callbacks/promise syntax

File: package.json (hunk 2 above)
Comment (DaveLak) [reply package.json]: Ah right, and it needs a URI I think. Alrighty, I follow.

File: babel.config.js (hunk 4 above)
Comment (tday) [reply babel.config.js]: this is just used for tests right now. It will need to be updated for bundling for the browser

File: src/attribution.js [hunk 6]
```diff
@@ -75,6 +75,11 @@ function getUtmQueryString() {
   return undefined;
 }
 
+function getCookieSafeQueryString(unsafeQueryString) {
+  // cookie values must not contain '='
+  return unsafeQueryString.toLowerCase().replace(/=/g, '-->>');
```
Comment (DaveLak) [inline src/attribution.js]: Do we not URL encode these for readability? We're not worried about other characters (e.g. `?`, `:`, `@`, `&`, `+`, `$`, `#`)? If we are maybe we should use `encodeURIComponent()` instead?

Comment (tday) [reply src/attribution.js]: Changing this behavior has a larger impact so I'd rather not change it for now.

Comment (DaveLak) [reply src/attribution.js]: Makes sense

**Pull request merged**